contracts/src/L1/rollup/ScrollChain.sol

```diff
@@ -441,56 +573,195 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
+                _skippedL1MessageBitmap
+            );
+            unchecked {
+                _totalL1MessagesPoppedInBatch += _totalNumL1MessagesInChunk;
+                _totalL1MessagesPoppedOverall += _totalNumL1MessagesInChunk;
+            }
+            assembly {
+                mstore(batchDataHashPtr, _chunkDataHash)
+                batchDataHashPtr := add(batchDataHashPtr, 0x20)
+            }
+        }
+
+        // compute the data hash for current batch
+        assembly {
+            let dataLen := mul(_chunksLength, 0x20)
+            _batchDataHash := keccak256(sub(batchDataHashPtr, dataLen), dataLen)
+        }
+    }
+
     /// @dev Internal function to load batch header from calldata to memory.
     /// @param _batchHeader The batch header in calldata.
-    /// @return memPtr The start memory offset of loaded batch header.
+    /// @return batchPtr The start memory offset of loaded batch header.
     /// @return _batchHash The hash of the loaded batch header.
-    function _loadBatchHeader(bytes calldata _batchHeader) internal pure returns (uint256 memPtr, bytes32 _batchHash) {
-        // load to memory
-        uint256 _length;
-        (memPtr, _length) = BatchHeaderV0Codec.loadAndValidate(_batchHeader);
+    /// @return _batchIndex The index of this batch.
+    /// @param _totalL1MessagesPoppedOverall The number of L1 messages popped after this batch.
+    function _loadBatchHeader(bytes calldata _batchHeader)
+        internal
+        view
+        returns (
+            uint256 batchPtr,
+            bytes32 _batchHash,
+            uint256 _batchIndex,
+            uint256 _totalL1MessagesPoppedOverall
+        )
+    {
+        // load version from batch header, it is always the first byte.
+        uint256 version;
+        assembly {
+            version := shr(248, calldataload(_batchHeader.offset))
+        }
 
-        // compute batch hash
-        _batchHash = BatchHeaderV0Codec.computeBatchHash(memPtr, _length);
+        // version should be always 0 or 1 in current code
+        uint256 _length;
+        if (version == 0) {
+            (batchPtr, _length) = BatchHeaderV0Codec.loadAndValidate(_batchHeader);
+            _batchHash = BatchHeaderV0Codec.computeBatchHash(batchPtr, _length);
+            _batchIndex = BatchHeaderV0Codec.batchIndex(batchPtr);
+        } else if (version == 1) {
+            (batchPtr, _length) = BatchHeaderV1Codec.loadAndValidate(_batchHeader);
+            _batchHash = BatchHeaderV1Codec.computeBatchHash(batchPtr, _length);
+            _batchIndex = BatchHeaderV1Codec.batchIndex(batchPtr);
+        } else {
+            revert ErrorInvalidBatchHeaderVersion();
+        }
+        // only check when genesis is imported
+        if (committedBatches[_batchIndex] != _batchHash && finalizedStateRoots[0] != bytes32(0)) {
+            revert ErrorIncorrectBatchHash();
+        }
+        _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.totalL1MessagePopped(batchPtr);
```

Comment:
what about putting this into `if-else`? e.g., to make the code style more consistent (although the logic is the same).

Reply:
the if is only for `revert`.

Reply:
oh, I mean:
```
if (_version == 0) {
  ...
  _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
} else {_version == 1} {
  ...
  _totalL1MessagesPoppedOverall = BatchHeaderV1Codec.getTotalL1MessagePopped(batchPtr);
}
```
but yeah, no need to differentiate this, because I found `BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);` is also used in many common places, like:
```
        // Pop finalized and non-skipped message from L1MessageQueue.
        _popL1Messages(
            BatchHeaderV0Codec.getSkippedBitmapPtr(memPtr),
            BatchHeaderV0Codec.getTotalL1MessagePopped(memPtr),
            BatchHeaderV0Codec.getL1MessagePopped(memPtr)
        );
```